packages/next/src/server/app-render/dynamic-rendering.ts

```diff
@@ -655,39 +654,52 @@ function createErrorWithComponentStack(
   return error
 }
 
-export function throwIfDisallowedDynamic(
+export function throwIfDisallowedEmptyShell(
   route: string,
   dynamicValidation: DynamicValidationState,
   serverDynamic: DynamicTrackingState,
   clientDynamic: DynamicTrackingState
 ): void {
-  let syncError: null | Error
-  let syncExpression: undefined | string
-  let syncLogged: boolean
+  // TODO: Now that metadata is streaming we don't really need to check if it is blocking
+  // the root. We leave this here for now to ensure we've actually covered all our bases here
+  // but we can actually remove this in a future update
+  if (dynamicValidation.hasDynamicMetadata) {
+    throw new InvariantError(
+      'Expected `generateMetadata` not to block the application shell but it did.'
+    )
+  }
+
+  if (dynamicValidation.hasSuspenseAboveBody) {
+    // This route has opted into allowing fully dynamic rendering
+    // by including a Suspense boundary above the body. In this case
+    // a lack of a shell is not considered disallowed so we simply return
+    return
+  }
+
   if (serverDynamic.syncDynamicErrorWithStack) {
-    syncError = serverDynamic.syncDynamicErrorWithStack
-    syncExpression = serverDynamic.syncDynamicExpression!
-    syncLogged = serverDynamic.syncDynamicLogged === true
-  } else if (clientDynamic.syncDynamicErrorWithStack) {
-    syncError = clientDynamic.syncDynamicErrorWithStack
-    syncExpression = clientDynamic.syncDynamicExpression!
-    syncLogged = clientDynamic.syncDynamicLogged === true
-  } else {
-    syncError = null
-    syncExpression = undefined
-    syncLogged = false
+    // There is no shell and the server did something sync dynamic likely
+    // leading to an early termination of the prerender before the shell
+    // could be completed.
+    if (!serverDynamic.syncDynamicLogged) {
```

Comment:
If I'm not mistaken, this condition is never false, i.e. `serverDynamic.syncDynamicLogged` is never true. I don't see any code that sets `prerenderStore.validating` to `true`, which in turn would lead to `abortAndThrowOnSynchronousRequestDataAccess` setting `dynamicTracking.syncDynamicLogged` to `true`. Is this just dead code that can be removed, or is `prerenderStore.validating` supposed to be set in `spawnDynamicValidationInDev`?

Reply:
Well its not well factored but it is either undefined or true today. I could probably drop the validating prop and just infer from whether there was a prerender during dev and assume it was logged. But as it stands in prod this will always be undefined and thus this condition truthy and in dev it will always be true and thus this condition will be false